Sofa/Component/Engine/Select/src/sofa/component/engine/select/BaseROI.h

```diff
@@ -0,0 +1,172 @@
+*                                                                             *
+* This program is free software; you can redistribute it and/or modify it     *
+* under the terms of the GNU Lesser General Public License as published by    *
+* the Free Software Foundation; either version 2.1 of the License, or (at     *
+* your option) any later version.                                             *
+*                                                                             *
+* This program is distributed in the hope that it will be useful, but WITHOUT *
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or       *
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License *
+* for more details.                                                           *
+*                                                                             *
+* You should have received a copy of the GNU Lesser General Public License    *
+* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#pragma once
+#include <sofa/component/engine/select/config.h>
+
+#include <sofa/type/Vec.h>
+#include <sofa/core/DataEngine.h>
+#include <sofa/core/objectmodel/BaseObject.h>
+#include <sofa/core/behavior/MechanicalState.h>
+#include <sofa/core/topology/BaseMeshTopology.h>
+#include <sofa/defaulttype/RigidTypes.h>
+
+namespace sofa::component::engine::select
+{
+
+/**
+ * This interface defines ROI features.
+ */
+template <class DataTypes>
+class BaseROI : public core::DataEngine
+{
+public:
+    using Inherit = core::DataEngine;
+    SOFA_CLASS(SOFA_TEMPLATE(BaseROI,DataTypes), Inherit);
+
+    using VecCoord = typename DataTypes::VecCoord;
+    using Coord = typename DataTypes::Coord;
+    using Real = typename DataTypes::Real;
+    using CPos = typename DataTypes::CPos;
+
+    using PointID = core::topology::BaseMeshTopology::PointID;
+    using SetIndex = core::topology::BaseMeshTopology::SetIndex;
+    using Edge = core::topology::BaseMeshTopology::Edge;
+    using VecEdge = type::vector<Edge>;
+    using Triangle = core::topology::BaseMeshTopology::Triangle;
+    using VecTriangle = type::vector<Triangle>;
+    using Quad = core::topology::BaseMeshTopology::Quad;
+    using VecQuad = type::vector<Quad>;
+    using Tetra = core::topology::BaseMeshTopology::Tetra;
+    using VecTetra = type::vector<Tetra>;
+    using Hexa = core::topology::BaseMeshTopology::Hexa;
+    using VecHexa = type::vector<Hexa>;
+
+public:
+    void init() final;
+    void doUpdate() override; //cannot set final because of the tests;
```

Comment:
Can the test be changed to make `doUpdate` final?

Reply:
The aforementioned test is DataEngine_test, which inherits DataEngine and counting update numbers (very useful test...). It would need some refactoring to change this behavior unfortunately